src/call.cpp

```diff
@@ -123,201 +123,50 @@ unsigned int call::wake()
-    return end[0] == '\0' ? NULL : end;
-}
-
-/******* Media information management *************************/
-/*
- * Look for "c=IN IP4 " pattern in the message and extract the following value
- * which should be IP address
- */
-uint32_t get_remote_ip_media(char *msg)
-{
-    char pattern[] = "c=IN IP4 ";
-    char *begin, *end;
-    char ip[32];
-    char *my_msg = strdup(msg);
-
-    if (!my_msg) {
-        return INADDR_NONE;
-    }
-    begin = strstr(my_msg, pattern);
-    if (!begin) {
-        free(my_msg);
-        /* Can't find what we're looking at -> return no address */
-        return INADDR_NONE;
-    }
-    begin += sizeof(pattern) - 1;
-    end = find_sdp_eol(begin);
-    if (!end) {
-        free(my_msg);
-        return INADDR_NONE;
-    }
-    *end = '\0';
-    memset(ip, 0, 32);
-    strncpy(ip, begin, sizeof(ip) - 1);
-    ip[sizeof(ip) - 1] = '\0';
-    free(my_msg);
-    return inet_addr(ip);
-}
-
-/*
- * Look for "c=IN IP6 " pattern in the message and extract the following value
- * which should be IPv6 address
- */
-uint8_t get_remote_ipv6_media(char *msg, struct in6_addr *addr)
-{
-    char pattern[] = "c=IN IP6 ";
-    char *begin, *end;
-    char ip[128];
-    char *my_msg = strdup(msg);
-
-    memset(addr, 0, sizeof(*addr));
-    memset(ip, 0, 128);
+    std::string::size_type begin, end;
 
-    if (!my_msg) {
-        return 0;
-    }
-    begin = strstr(my_msg, pattern);
-    if (!begin) {
-        free(my_msg);
-        /* Can't find what we're looking at -> return no address */
-        return 0;
+    begin = msg.find(pattern);
+    if (begin == std::string::npos) {
+        return "";
     }
-    begin += sizeof(pattern) - 1;
-    end = find_sdp_eol(begin);
-    if (!end) {
-        free(my_msg);
-        return 0;
-    }
-    *end = '\0';
-    strncpy(ip, begin, sizeof(ip) -1);
-    ip[sizeof(ip) - 1] = '\0';
-    free(my_msg);
-    if (!inet_pton(AF_INET6, ip, addr)) {
-        return 0;
+
+    begin += pattern.size();
+    end = msg.find_first_of(" \r\n", begin);
+    if (end == std::string::npos || begin == end) {
+        return "";
     }
-    return 1;
+
+    return msg.substr(begin, end - begin);
 }
 
-/*
- * Look for "m=audio ", "m=image " or "m=video " pattern in the message
- * and extract the following value which should be port number.
- */
-enum media_ptn {
-    PAT_AUDIO,
-    PAT_IMAGE,
-    PAT_VIDEO
-};
-uint16_t get_remote_port_media(const char *msg, enum media_ptn pattype)
+void call::get_remote_media_addr(std::string const &msg)
 {
-    const char *pattern;
-    char *begin, *end;
-    char number[6];
-
-    if (pattype == PAT_AUDIO) {
-        pattern = "m=audio ";
-    } else if (pattype == PAT_IMAGE) {
-        pattern = "m=image ";
-    } else if (pattype == PAT_VIDEO) {
-        pattern = "m=video ";
-    } else {
-        ERROR("Internal error: Undefined media pattern %d\n", 3);
+    std::string port, host = parse_sdp_msg(media_ip_is_ipv6 ? "c=IN IP6 " : "c=IN IP4 ", msg);
```

Comment:
Declaring the port here is confusing, looks almost like parse_sdp_msg returns a tuple (with my python eyes).